### PR TITLE
errors: helper for parameter-related errors

### PR DESCRIPTION
**Pull request by rwcarlsen**
Provides a helper function ``paramError([param-name], [msg])`` for emitting moose errors related to input parameters.  The error will automatically be formatted with a prefix that includes the full parameter path (including the object name) and the input file name and line number of corresponding to the bad value.

For example:

```c++
// ... inside a moose object constructor:
if (getParam<int>("myparam") <= 42)
  paramError("myparam", "value must be greater than 42");
```

with this input file:

```
[Foo]
  [bar]
    myparam = 41
  []
[]
```

gives this error:

```
*** ERROR ***
/path/to/infile.i:3: (Foo/bar/myparam) value must be greater than 42
```

I plan on working converting more moose error's to use paramError off and on in coming weeks.  Others are welcome to as well.

### CODE REVIEW

**Comment by permcody** [comment]
A couple of questions: 
 1. How does `paramError` handle parameters that don't come through the input file (i.e. object built through an Action)?
 2. Does paramError do the right thing with GlobalParams? I assume it just reports the value from GlobalParams which is useful, what about an input file that has the value in the block AND in GlobalParams?

**Comment by rwcarlsen** [comment]
1. It falls back to the regular mooseError behavior.
2. Yes - it returns ``GlobalParams/foo`` in the prefix correctly with the correct line.  If both global and local are present, the local one takes precedence just like things always have.

File: framework/src/bcs/IntegratedBC.C [hunk 1]
```diff
@@ -104,9 +104,10 @@ IntegratedBC::IntegratedBC(const InputParameters & parameters)
     MooseVariable * var = &_subproblem.getVariable(_tid, _diag_save_in_strings[i]);
 
     if (var->feType() != _var.feType())
-      mooseError("Error in " + name() + ". When saving diagonal Jacobian values in an Auxiliary "
-                                        "variable the AuxVariable must be the same type as the "
-                                        "nonlinear variable the object is acting on.");
+      paramError(
+          "diag_save_in",
+          "saved-in auxiliary variable is incompatible with the object's nonlinear variable: ",
+          moose::internal::incompatVarMsg(*var, _var));
```
Comment (permcody) [inline framework/src/bcs/IntegratedBC.C]: Does this method really belong in the moose::internal namespace? It looks like it's something that should be exposed to end users.

File: framework/include/base/MooseObject.h [hunk 2]
```diff
@@ -76,6 +77,20 @@ class MooseObject : public ConsoleStreamInterface, public libMesh::ParallelObjec
    */
   virtual bool enabled() { return _enabled; }
 
+  /**
+   * Emits an error prefixed with the file and line number of the given param (from the input
+   * file) with the given args as the message.
+   */
+  template <typename... Args>
+  [[noreturn]] void paramError(std::string param, Args... args)
```
Comment (permcody) [inline framework/include/base/MooseObject.h]: Prefer `const std::string &`  here and in other methods.

File: test/tests/misc/save_in/tests [hunk 3]
```diff
@@ -10,12 +10,12 @@
...
     type = RunException
     input = 'save_in_soln_var_err_test.i'
-    expect_err = 'Trying to use solution variable u as a save_in variable in diff'
+    expect_err = 'diff/save_in\) cannot use solution variable as save-in variable'
   [../]
   [./test_diag_sol_var_err]
     type = RunException
     input = 'diag_save_in_soln_var_err_test.i'
-    expect_err = 'Trying to use solution variable u as a diag_save_in variable in diff'
+    expect_err = 'diff/diag_save_in\) cannot use solution variable as diag save-in variable'
```
Comment (permcody) [inline test/tests/misc/save_in/tests]: This is OK, but it's like a scratch you can't itch (unbalanced parenthesis)! 
Can you change the match to \(.*diff/diag_save_in\)? Actually we should probably decide if this is the format we want (input syntax showing up in parenthesis). Maybe we should use brackets or instead since that's the symbol in the input file? In either case, I'd prefer to see more matched here.

File: framework/src/bcs/IntegratedBC.C (hunk 1 above)
Comment (permcody) [reply framework/src/bcs/IntegratedBC.C]: BTW - I'm supportive of migrating to better name spacing in the long run. We'll have to come up with a way to support the transition. Perhaps we can do this by creating the new namespace with the desired declarations and calling methods in the existing namespace?

File: framework/include/base/MooseError.h [hunk 4]
```diff
@@ -104,11 +104,15 @@
...
 [[noreturn]] void mooseError(Args &&... args);
 
+class MooseVariable;
+
 namespace moose
 {
 namespace internal
 {
 
+std::string incompatVarMsg(MooseVariable & var1, MooseVariable & var2);
```
Comment (permcody) [inline framework/include/base/MooseError.h]: Need a really good documentation string for this method. It's unclear what this method does from the name. In fact, I don't think it really does what the name implies (prints an incompatible variable message).

File: framework/include/base/MooseObject.h [hunk 5]
```diff
@@ -18,6 +18,7 @@
 // MOOSE includes
 #include "InputParameters.h"
 #include "ConsoleStreamInterface.h"
+#include "Parser.h"
```
Comment (permcody) [inline framework/include/base/MooseObject.h]: Oh boy... The Parser makes inroads further into MOOSE! This is as core as it gets.

**Review by permcody — CHANGES_REQUESTED** [review]
Overall this looks Ok. We need better documentation and I have a few other misc. comments. I'll probably wait to merge this until the team is back. This is a big enough change that I'll want to chat it over with the team. It ties us much more closely to the parser than we've been in the passed, but does indeed give us nice messages.

File: framework/include/base/MooseObject.h (hunk 5 above)
Comment (rwcarlsen) [reply framework/include/base/MooseObject.h]: This actually doesn't have to go here.  We could move the special info param name functions (paramLocName, paramPathName) into MooseObject.h/C which Parser.C already has access to.

File: framework/src/bcs/IntegratedBC.C (hunk 1 above)
Comment (rwcarlsen) [reply framework/src/bcs/IntegratedBC.C]: I don't anticipate users often writing code that needs to check for moosevariable type compatibility.  And even if they do, I think the chances that they know about and use this method (or that we notice and inform them of it) are pretty low.  If we see specific cases where those assumptions begin to be violated, it is easy to pull it out of "internal", but not as much the other way around.

File: test/tests/misc/save_in/tests (hunk 3 above)
Comment (rwcarlsen) [reply test/tests/misc/save_in/tests]: Brackets would be a bit weird because the actual parameter names are not part of the block/section names.  The actual hit equivalent syntax for the parameter is without brackets or parens (i.e. ``foo/bar=baz``).

**Comment by permcody** [comment]
@idaholab/moose-team - This PR adds a new core error type to MOOSE "ParamError", that will automatically print line number information along with the error when that parameter is set from an input file. I want to call attention to this PR to make sure that people are happy with the interface and method name before we commit. This ties the syntax a little closer to MOOSE, but I think it's a step forward in providing more useful error reporting. Comments?

**Comment by rwcarlsen** [comment]
Note that ``paramError` is a member function of ``MooseObject`` - it does not have a globally scoped function version like ``mooseError`` does.

File: framework/include/base/MooseError.h [hunk 6]
```diff
@@ -104,11 +104,21 @@
...
 {
 namespace internal
 {
 
+/// Builds and returns a string of the form:
+///
+///     [var1-elemtype],ORDER[var1-order] != [var2-elemtype],ORDER[var2-order]
+///
+/// This is a convenience function to be used when error messages (especially with paramError)
+/// need to report that variable types are incompatible (e.g. with residual save-in).
```
Comment (dschwen) [inline framework/include/base/MooseError.h]: comment style...

Comment (rwcarlsen) [reply framework/include/base/MooseError.h]: The entire file is that way.

Comment (dschwen) [reply framework/include/base/MooseError.h]: Booooo

**Comment by aeslaughter** [comment]
Having line numbers on parameter related errors is great, whatever it takes I like it.

**Comment by permcody** [comment]
OK - well I have no major issues. Just wondering if we all agree on the new API, "ParamError". We have MooseAssert, MooseError, MooseWarning, MooseDeprecated, and now "ParamError", (MooseParamError?).

**Comment by rwcarlsen** [comment]
All the others have global variants.  paramError does not because it
can't.  I don't think it would make sense to try too hard to follow the
"moose"-prefix convention; it is only a MooseObject member function.

**Comment by permcody** [comment]
Good Point, I'm satisfied

**Review by permcody — APPROVED** [review]
